### PR TITLE
feat: Query Lend APY

### PR DESCRIPTION
**Pull request by RafilxTenfen**
## Description

- Add structures to retrieve the Lend APY of a specific denom

progress toward #343 

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/leverage/types/keys.go [hunk 1]
```diff
@@ -29,6 +29,8 @@ var (
 	KeyPrefixLastInterestTime  = []byte{0x06}
 	KeyPrefixExchangeRate      = []byte{0x07}
 	KeyPrefixBadDebt           = []byte{0x08}
+	KeyPrefixBorrowAPY         = []byte{0x09}
+	KeyPrefixLendAPY           = []byte{0x10}
```
Comment (toteki) [inline x/leverage/types/keys.go:33]:
```suggestion
	KeyPrefixLendAPY           = []byte{0x0A}
```

File: x/leverage/keeper/lend.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+func (k Keeper) SetLendAPY(ctx sdk.Context, denom string, lendAPY sdk.Dec) error {
+	store := ctx.KVStore(k.storeKey)
+	key := types.CreateLendAPYKey(denom)
+
+	bytesAPY, err := lendAPY.Marshal()
+	if err != nil {
+		return err
+	}
+
+	store.Set(key, bytesAPY)
```
Comment (toteki) [inline x/leverage/keeper/lend.go:39]:
```suggestion
	bz, err := lendAPY.Marshal()
	if err != nil {
		return err
	}

	store.Set(key, bz)
```

File: x/leverage/keeper/lend.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/umee-network/umee/x/leverage/types"
+)
+
+// GetLendAPY returns an sdk.Dec of an lend APY
+// returns sdk.ZeroDec if not found
+func (k Keeper) GetLendAPY(ctx sdk.Context, denom string) sdk.Dec {
+	store := ctx.KVStore(k.storeKey)
+	key := types.CreateLendAPYKey(denom)
+
+	bytesAPY := store.Get(key)
```
Comment (toteki) [inline x/leverage/keeper/lend.go:15]: apply `bz` convention

Comment (RafilxTenfen) [reply x/leverage/keeper/lend.go]: done at https://github.com/umee-network/umee/pull/367/commits/a8d6e721cebcee2d31aac58c5ff938e00d415087

**Review by toteki — APPROVED** [review]

File: x/leverage/keeper/interest.go [hunk 4]
```diff
@@ -101,6 +101,10 @@ func (k Keeper) AccrueAllInterest(ctx sdk.Context) error {
 		if err := k.SetBorrowAPY(ctx, coin.Denom, derivedRate); err != nil {
 			return err
 		}
+
+		if err := k.SetLendAPY(ctx, coin.Denom, derivedRate.Mul(utilization)); err != nil {
```
Comment (alexanderbez) [inline x/leverage/keeper/interest.go:105]: @toteki can you remind me what `utilization` is here again?

nit: I think we should call the `derivedRate` variable `borrowRate` instead.

File: x/leverage/keeper/lend.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/umee-network/umee/x/leverage/types"
+)
+
+// GetLendAPY returns an sdk.Dec of an lend APY
+// returns sdk.ZeroDec if not found
```
Comment (alexanderbez) [inline x/leverage/keeper/lend.go:11]:
```suggestion
// GetLendAPY returns an sdk.Dec of an lend APY returns sdk.ZeroDec if not found.
```

File: x/leverage/keeper/lend.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+	var lendAPY sdk.Dec
+	if err := lendAPY.Unmarshal(bz); err != nil {
+		panic(err)
+	}
+
+	return lendAPY
+}
+
+// SetLendAPY sets the lend APY of an specific denom
```
Comment (alexanderbez) [inline x/leverage/keeper/lend.go:29]:
```suggestion
// SetLendAPY sets the lend APY of an specific denom.
```

**Review by alexanderbez — APPROVED** [review]

File: x/leverage/keeper/interest.go (hunk 4 above)
Comment (toteki) [reply x/leverage/keeper/interest.go:105]: Borrow utilization is an `sdk.Dec` ranging from 0 to 1, which represents the portion of a given lent asset that has been borrowed.

For example if lenders have lent a total of 22000 uatom (excluding reserves) and borrowers have borrowed 11000 (after interest), then uatom is at 50% borrow utilization.

**Pull request merged**